### PR TITLE
Add reference count to LibGit2 objects

### PR DESCRIPTION
**Pull request by simonbyrne**
Fixes segfault from calling `git_libgit2_shutdown` before finalizers (#20109).

### CODE REVIEW

**Comment by simonbyrne** [comment]
I just realised that one alternative to this would be to call `git_libgit2_init` on each object initialization, and `git_libgit2_shutdown` at each finalizer, as these act like a reference counter.

**Comment by kshyatt** [comment]
Can this be reviewed/merged?

**Comment by tkelman** [comment]
do we want the alternative, or this? pros/cons?

File: base/libgit2/types.jl [hunk 1]
```diff
@@ -490,6 +494,11 @@ for (typ, reporef, sup, cname) in [
         if obj.ptr != C_NULL
             ccall(($(string(cname, :_free)), :libgit2), Void, (Ptr{Void},), obj.ptr)
             obj.ptr = C_NULL
+            REFCOUNT[] -= 1
+            if REFCOUNT[] == 0
+                # will the last finalizer please turn out the lights?
+                ccall((:git_libgit2_shutdown, :libgit2), Cint, ())
```
Comment (Keno) [inline base/libgit2/types.jl]: What if the number of libgit2 objects drops to 0 at runtime. Is that ok?

File: base/libgit2/libgit2.jl [hunk 2]
```diff
@@ -9,6 +9,8 @@ export with, GitRepo, GitConfig
 const GITHUB_REGEX =
     r"^(?:git@|git://|https://(?:[\w\.\+\-]+@)?)github.com[:/](([^/].+)/(.+?))(?:\.git)?$"i
 
+const REFCOUNT = Ref(zero(UInt))
```
Comment (yuyichao) [inline base/libgit2/libgit2.jl]: I would set this to `0` in `__init__` also.

File: base/libgit2/libgit2.jl [hunk 3]
```diff
@@ -603,8 +605,14 @@ function __init__()
 
     err = ccall((:git_libgit2_init, :libgit2), Cint, ())
     err > 0 || throw(ErrorException("error initializing LibGit2 module"))
+    REFCOUNT[] += 1
```
Comment (yuyichao) [inline base/libgit2/libgit2.jl]: Ideally we'll need to replace these with atomic counters but this should be good enough for now.

**Review by yuyichao — CHANGES_REQUESTED** [review]

File: base/libgit2/libgit2.jl (hunk 2 above)
Comment (yuyichao) [inline base/libgit2/libgit2.jl]: `s/0/1/g`

File: base/libgit2/types.jl (hunk 1 above)
Comment (yuyichao) [inline base/libgit2/types.jl]: The `__init__` holds a global reference so the reference should not drop to 0.

**Comment by simonbyrne** [comment]
> do we want the alternative, or this? pros/cons?

The only differences I can tell:
- the alternative seems to [have some code for handling threading](https://github.com/libgit2/libgit2/blob/5fe5557e8a8d3fd6a4617c2d8c863c1f62848020/src/global.c#L153-L162) (I don't know how this would perform, though @yuyichao mentions atomic counter above
- the alternative uses a `Cint` for the counter, so there is theoretically a potential for overflow on 64-bit machines.

File: base/libgit2/libgit2.jl (hunk 2 above)
Comment (simonbyrne) [inline base/libgit2/libgit2.jl]: done.

File: base/libgit2/libgit2.jl (hunk 3 above)
Comment (simonbyrne) [inline base/libgit2/libgit2.jl]: I've changed these to be atomic counters: can you have a quick look over it as I'm not really familiar with them.

**Comment by kshyatt** [comment]
Did the changes @yuyichao want get made? Can we merge this?

**Review by yuyichao — APPROVED** [review]

**Pull request merged**